### PR TITLE
fix 38503.

### LINKED ISSUE #38503: Fail to short circuit an undefined private class field with optional chaining
**TypeScript Version:**  3.8.3

**Search Terms:**
* private method
* private class field
* short circuit
* optional chaining
* typeerror
* cannot get property of

**Code**

```ts
     /** Truncated for brevity */
33   // this.#callback?.(this.#list, this); /** Fails to short-circuit with ?. */
34   if (this.#callback) this.#callback(this.#list, this); /** This works fine */
     /** Truncated for brevity */
```

**Expected behavior:**

Expected `this.#callback?.(this.#list, this)` to not execute when `this.#callback` is nullish.

**Actual behavior:**

```ts
/**
 * Fail to short-circuit with ?. and produces TypeError.
 * Expected it to behave just like:
 * 
 *   if (this.#callback) this.#callback(this.#list, this);
 *
 */
this.#callback?.(this.#list, this);
```

```sh
Uncaught TypeError: Cannot read property 'call' of undefined
    at A.add (eval at <anonymous> (main-3.js:1239), <anonymous>:42:62)
    at eval (eval at <anonymous> (main-3.js:1239), <anonymous>:54:3)
    at main-3.js:1239
```

**Playground Link:** 

Run the [reduced test case in TypeScript playground] and see the result in DevTools.
Comment and uncomment L33 and L34 to see the different results when they should behave the same.

**Related Issues:** 
Tried to look up but could not find one.

[reduced test case in TypeScript playground]: https://www.typescriptlang.org/play/?ssl=19&ssc=2&pln=20&pc=1#code/MYGwhgzhAECCAyBLCAXaBvAUNH0DEIyKAXNAHYCuAtgEYCmATgNoC60AvNKwNyba4AHCjULBoYACYSAFAA9SlWowCUpAG4B7RBIz9cuFAAtkAOgJETQiIbnLe+gL599QkYjGg6YBtNXRN2rr6+kam5qgcXCz2uE560K6i0HRkKAyIdBC+CtT0zGxYwbgMdCgUDGTQoRBmhKgxOE5xiKmMAGZgwHRwAAoMGgKMKBkwhbjAYCAgNJ0A1tJ1JHBIqAA00Bo0AFZ0wEuwfgESvHGgkDCwQbjh+ytonGR0AO7LRL4N+BNTM8CzAPykWB9AZDEZMADkX2mc3B0WcgmESWAGjIqAYFD2Gh8UJ+-0BwMGDGGmQhOJhLGUVyKiDa0GkKAAnoMNLTpGTfpT2FzoOC2hQyHtECjwZTqmZ2bNIhKPnEXIj3OIpHIckoGIctDoxkUxTcTJIZLI7PCigB6ABUZugACEAKoAcVIABVjDAJhQIJlkgx+gxoGaTfF9CaTVUXeLJtDfn8TPSwzd1tUjUUcDS6TqJaK4xLY2FFgmXUnYsaEvKPCAvD51YEtcEdYsTJ5vO94k0+MjUWgwJFHi9YM32xEaN3nnBpAsiOtNhBOQA+KnQAcacsmEAaADm0kDuAAsmAjCY2qusdJCW0sVQwAK6CYyBonr4-dAAIwABmgAFJny-lKstzgp7+ybQPWKRpCMvjxEaDhGpgYB6kqT5GnB+rSAATDBNDwTIADMRqYShAAsdhAA

### PR DESCRIPTION
**Pull request by ShuiRuTian**
Fixes #38503

### CODE REVIEW

**Comment by ShuiRuTian** [comment]
```
##[error]node_modules/@octokit/plugin-rest-endpoint-methods/dist-types/generated/parameters-and-response-types.d.ts(961,60): error TS2339: Property 'GET /user/:migration_id/repositories' does not exist on type 'Endpoints'.
##[error]node_modules/@octokit/plugin-rest-endpoint-methods/dist-types/generated/parameters-and-response-types.d.ts(962,33): error TS2339: Property 'GET /user/:migration_id/repositories' does not exist on type 'Endpoints'.
##[error]node_modules/@octokit/plugin-paginate-rest/dist-types/generated/paginating-endpoints.d.ts(987,31): error TS2339: Property 'GET /user/:migration_id/repositories' does not exist on type 'Endpoints'.
##[error]node_modules/@octokit/plugin-paginate-rest/dist-types/generated/paginating-endpoints.d.ts(988,29): error TS2339: Property 'GET /user/:migration_id/repositories' does not exist on type 'Endpoints'.
```
some wrong with the test?
Is it due to mine change or something else like net connection?

**Comment by orta** [comment]
this should be fixed: https://github.com/microsoft/TypeScript/issues/39712

**Comment by ShuiRuTian** [comment]
Friendly ping ~ @rbuckton

**Comment by ShuiRuTian** [comment]
ping @rbuckton

**Comment by ShuiRuTian** [comment]
up.

File: tests/baselines/reference/privateNameFieldCallExpression.js [hunk 1]
```diff
@@ -35,15 +36,17 @@ class A {
     }
     test() {
         var _a;
+        var _b;
         __classPrivateFieldGet(this, _fieldFunc).call(this);
+        (_a = __classPrivateFieldGet(this, _fieldFunc)) === null || _a === void 0 ? void 0 : _a(this);
```
Comment (rbuckton) [inline tests/baselines/reference/privateNameFieldCallExpression.js:41]: I don't think this fixes the issue. The expectation is that `this.#fieldFunc` will be called with the correct `this`, but instead its called with `this` as an argument. In other words, it should read like this instead:

```js
(_a = __classPrivateFieldGet(this, _fieldFunc)) === null || _a === void 0 ? void 0 : _a.call(this);
```

File: src/compiler/transformers/classFields.ts [hunk 2]
```diff
@@ -341,7 +341,8 @@ namespace ts {
                 const { thisArg, target } = factory.createCallBinding(node.expression, hoistVariableDeclaration, languageVersion);
                 return factory.updateCallExpression(
                     node,
-                    factory.createPropertyAccessExpression(visitNode(target, visitor), "call"),
+                    // #38503
```
Comment (rbuckton) [inline src/compiler/transformers/classFields.ts:344]: Unfortunately, this isn't enough to get the correct emit, but its the right start. I tinkered with your PR locally and I think I have a solution that builds off of what you've done here. Essentially, you need to move the `isCallChain` test out of the call to `updateCallExpression` and call `updateCallChain` in its own branch. In the new `updateCallChain` branch, you "move" the `?.` from the original call (i.e., `this.#x?.()`) to the new property access for `.call` needed to preserve the correct `this` (i.e., `_a?.call(this)`).

I can see if I can push an update to your PR with this change shortly.

**Review by rbuckton — CHANGES_REQUESTED** [review]

**Review by rbuckton — APPROVED** [review]

**Comment by rbuckton** [comment]
Thanks for the contribution!

**Pull request merged**